### ISSUE TITLE
Strict equality false positive when strict optional disabled

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#8296]
This generates an unexpected error:

```py
# mypy: no-strict-optional, strict-equality

from typing import Any, Optional

x: Optional[Any]

if x in (1, 2):  # Error: non-overlapping container check
    pass
```

With strict optional checking enabled, it works as expected and doesn't generate an error.
